File: src/runbook_voice/booking_bridge.py
"""The Track A <-> Track C seam: dispatch runbook steps to the JS booking modules.

``RunbookExecutor`` drives steps through ``PersistentSailboxRunner.execute``.
Every implementation of that protocol until now has been a test fake, so
``restaurant.book`` resolved to nothing and the warm path could not run end to
end. ``NodeBookingRunner`` is the production implementation.

One subprocess per step, talking JSON over stdin/stdout to
``scripts/booking_bridge.mjs``. A subprocess rather than a long-lived server
because a step is infrequent and slow (a human is speaking between them), and a
server would add a lifecycle to own and a port to collide on for no gain.

Confirmation: ``RunbookExecutor`` gates irreversible steps *before* dispatching
them, so by the time this runner is called for ``restaurant.book`` the spoken
confirmation has already succeeded. The bridge still refuses to book unless
told so explicitly, so that fact must be asserted at construction with
``confirmation_is_upstream=True``. It defaults to False: a runner built without
thinking about it cannot book.
"""

from __future__ import annotations

import asyncio
import json
import os
from collections.abc import Mapping
from pathlib import Path
from typing import Any

_BRIDGE = Path(__file__).resolve().parents[2] / "scripts" / "booking_bridge.mjs"
_IRREVERSIBLE = frozenset({"restaurant.book"})


class BookingBridgeError(RuntimeError):
    """The bridge refused, failed, or returned something unusable.

    Raised rather than returned so ``RunbookExecutor`` records it as a failed
    step and stops - a booking action that half-worked must never look like
    success to the voice layer.
    """


class NodeBookingRunner:
    """Dispatch runbook actions to the JavaScript booking modules."""

    def __init__(
        self,
        *,
        stub: bool = False,
        confirmation_is_upstream: bool = False,
        store_path: str | os.PathLike[str] | None = None,
        node: str = "node",
        script: str | os.PathLike[str] | None = None,
        timeout: float = 180.0,
        env: Mapping[str, str] | None = None,
    ) -> None:
        self._node = node
        self._script = Path(script) if script else _BRIDGE
        self._timeout = timeout
        self._confirmation_is_upstream = confirmation_is_upstream

        base = dict(env if env is not None else os.environ)
        if stub:
            base["BOOKING_STUB_MODE"] = "1"
        else:
            # Never inherit a stray stub flag into what is meant to be a real
            # run: a stub booking mistaken for a real one is the single most
            # damaging failure this component has.
            base.pop("BOOKING_STUB_MODE", None)
        if store_path is not None:
            base["BOOKING_STORE_PATH"] = str(store_path)
        self._env = base

    @property
    def stub(self) -> bool:
        return self._env.get("BOOKING_STUB_MODE") in {"1", "true"}

    async def execute(self, action: str, arguments: Mapping[str, Any]) -> Any:
        """Run one runbook step. Signature matches PersistentSailboxRunner."""
        request = {
            "action": action,
            "arguments": dict(arguments),
            # Only ever true for steps the executor already gated by voice.
            "confirmed": action in _IRREVERSIBLE and self._confirmation_is_upstream,
        }

        try:
            process = await asyncio.create_subprocess_exec(
                self._node,
                str(self._script),
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=self._env,
            )
        except FileNotFoundError as exc:
            raise BookingBridgeError(
                f"cannot launch the booking bridge: {self._node!r} not found"
            ) from exc

        try:
            stdout, stderr = await asyncio.wait_for(
                process.communicate(json.dumps(request).encode()), self._timeout
            )
        except TimeoutError:
            process.kill()
            await process.wait()
            raise BookingBridgeError(
                f"booking bridge timed out after {self._timeout}s on {action!r}"
            ) from None

        return self._parse(action, stdout, stderr)

    def _parse(self, action: str, stdout: bytes, stderr: bytes) -> Any:
        text = stdout.decode().strip()
        if not text:
            detail = stderr.decode().strip() or "no output"
            raise BookingBridgeError(f"booking bridge produced no response: {detail}")

        try:
            payload = json.loads(text.splitlines()[-1])
        except json.JSONDecodeError as exc:
            raise BookingBridgeError(
                f"booking bridge returned non-JSON: {text[:200]}"
            ) from exc

        if not payload.get("ok"):
            raise BookingBridgeError(
                f"{action}: {payload.get('error', 'unknown bridge error')}"
            )
        return payload.get("result")

File: scripts/booking_bridge.mjs
#!/usr/bin/env node
// Track A (Python) <-> Track C (JavaScript) seam.
//
// Reads ONE JSON request on stdin, writes ONE JSON response on stdout, exits.
// One process per step: the booking modules keep no in-memory state worth
// preserving between steps (the booking store is on disk), and a short-lived
// process cannot leak a half-open browser session between runs.
//
//   in : {"action": "...", "arguments": {...}, "confirmed": false}
//   out: {"ok": true,  "result": {...}}
//        {"ok": false, "error": "...", "type": "ErrorName"}
//
// CONFIRMATION POLICY - the important part.
// There are two confirmation gates in this system: RunbookExecutor._confirm()
// on the Python side (spoken, templated readback, exact-yes) and confirmGate()
// inside bookStep(). Invariant 1 wants exactly ONE decision point, and the
// spoken one is the real one - it is what a human actually hears and answers.
//
// So Python is authoritative and this bridge REFUSES any irreversible action
// unless the caller passes confirmed:true, which NodeBookingRunner only sends
// when it was constructed with confirmation_is_upstream=True. Absent that flag
// this endpoint cannot book, which keeps a direct/scripted caller fail-closed.
// bookStep's own gate is then satisfied with the decision already obtained
// rather than asking a second time - a second prompt after the user already
// said yes reads as a bug on stage and trains people to double-approve.

import { bookStep } from "../src/booking/book.js";
import { resetAll } from "../src/booking/resetScript.js";
import { listOpenBookings } from "../src/booking/store.js";
import { isStubMode } from "../src/booking/stubMode.js";
import { getProvider } from "../src/booking/providers/index.js";

const IRREVERSIBLE = new Set(["restaurant.book"]);

/** snake_case from the runbook -> camelCase the booking modules expect. */
function camel(key) {
  return key.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
}

/**
 * Map runbook arguments onto bookStep params.
 *
 * The runbook supplies {cuisine, city, date, time, party_size}; bookStep needs
 * {restaurant, date, time, partySize}. There is no `restaurant` slot because
 * the caller does not know one yet - picking it is the *search's* job.
 *
 * `restaurant` is therefore the provider's search term, and OpenTable's own
 * field is "location, restaurant, or cuisine", so a cuisine is a legitimate
 * value for it. We prefer an explicit `restaurant` when a synthesized runbook
 * provides one (TON-21 may lift a concrete name into a slot), and fall back to
 * cuisine. If neither exists we refuse rather than searching for "undefined".
 */
function toBookingParams(args) {
  const params = {};
  for (const [key, value] of Object.entries(args ?? {})) {
    params[camel(key)] = value;
  }
  if (!params.restaurant) {
    if (!params.cuisine) {
      throw new Error(
        "cannot determine a search term: runbook supplied neither `restaurant` nor `cuisine`"
      );
    }
    params.restaurant = params.cuisine;
  }
  if (params.partySize != null) params.partySize = String(params.partySize);
  return params;
}

/**
 * This process is either running on the laptop (default) or has been
 * re-invoked INSIDE the persistent "booking" Sailbox by delegateToSailbox()
 * below. Only the in-box copy can reach the browser: the Sailbox's Chromium
 * exposes its debug port on 127.0.0.1 only, deliberately never on a public
 * interface, because CDP has no built-in auth of its own - whoever can reach
 * it fully controls the browser, cookies and all. Tunneling that out to the
 * open internet (the way the VNC bridge did for a *password-protected*
 * viewer) is not a tradeoff worth making for this.
 */
const INSIDE_SAILBOX = process.env.BOOKING_BRIDGE_INSIDE_SAILBOX === "1";

/**
 * Reuses exactly one tab in the persistent session, closing any others.
 *
 * The box stays up across every rehearsal and every step within a run, and
 * nothing here ever closes a tab it opened - so stray tabs from a previous
 * step (or from interactive debugging against this same box) accumulate.
 * Found live: 5 open tabs, one on a chrome-error page, turned a 1s navigation
 * into a 30s timeout. Enforcing a single tab is cheap insurance against that
 * recurring during a rehearsal.
 */
async function acquirePage() {
  if (!INSIDE_SAILBOX) {
    throw new Error("acquirePage() must only run inside the Sailbox — see delegateToSailbox().");
  }
  const { chromium } = await import("playwright");
  const browser = await chromium.connectOverCDP("http://127.0.0.1:9222");
  const context = browser.contexts()[0];
  const pages = context.pages();
  const [keep, ...stray] = pages.length > 0 ? pages : [await context.newPage()];
  await Promise.all(stray.map((p) => p.close().catch(() => {})));
  return keep;
}

/**
 * Files the in-box copy of this bridge needs, mirrored under the same
 * relative layout (scripts/booking_bridge.mjs importing ../src/booking/*) so
 * its own `import` statements resolve unmodified once copied over.
 */
const SYNCED_FILES = [
  "scripts/booking_bridge.mjs",
  "src/booking/book.js",
  "src/booking/confirmGate.js",
  "src/booking/store.js",
  "src/booking/stubMode.js",
  "src/booking/guestProfile.js",
  "src/booking/paymentGuard.js",
  "src/booking/resetScript.js",
  "src/booking/providers/index.js",
  "src/booking/providers/opentable.js",
  "src/booking/providers/resy.js",
];
const REMOTE_ROOT = "/root/booking/repo";

async function syncToSailbox(box) {
  const { readFile } = await import("node:fs/promises");
  const path = await import("node:path");
  const { fileURLToPath } = await import("node:url");
  const localRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");

  for (const relative of SYNCED_FILES) {
    const contents = await readFile(path.join(localRoot, relative), "utf8");
    await box.fs.write(`${REMOTE_ROOT}/${relative}`, contents);
  }
}

/**
 * Re-invokes this same script inside the Sailbox with the original request on
 * stdin, over the Sail SDK's run() (no network port opened at all - the SDK
 * is the transport). Only reached for restaurant.search / restaurant.book in
 * real mode; stub mode and booking.reset/list_open never touch a page and
 * stay local.
 */
async function delegateToSailbox(request) {
  const { Sailbox } = await import("@sailresearch/sdk");
  const boxes = await Sailbox.list({ limit: 50 });
  const box = boxes.find((b) => (b.name ?? "") === "booking");
  if (!box) {
    throw new Error(
      'no Sailbox named "booking" found — has the TON-8 persistent box been created?'
    );
  }

  await syncToSailbox(box);
  await box.fs.write(`${REMOTE_ROOT}/request.json`, JSON.stringify(request));

  const storeEnv = process.env.BOOKING_STORE_PATH
    ? ` BOOKING_STORE_PATH=${JSON.stringify(process.env.BOOKING_STORE_PATH)}`
    : "";
  const command =
    `cd ${REMOTE_ROOT} && BOOKING_BRIDGE_INSIDE_SAILBOX=1${storeEnv} ` +
    `node scripts/booking_bridge.mjs < request.json`;
  const res = await box.run(command, { timeout: 120_000 });

  const lastLine = (res.stdout ?? "").trim().split("\n").filter(Boolean).pop();
  if (!lastLine) {
    throw new Error(`Sailbox bridge produced no output: ${(res.stderr ?? "").slice(0, 500)}`);
  }
  const payload = JSON.parse(lastLine);
  if (!payload.ok) throw new Error(payload.error);
  return payload.result;
}

/**
 * Resolve and VALIDATE the provider name, in stub mode too.
 *
 * getProvider() would catch an unknown name in real mode, but stub mode never
 * reaches it - so a runbook naming a provider that does not exist would pass
 * every rehearsal and throw on stage, at the irreversible step, in front of an
 * audience. Validating here means stub runs fail for the same reasons real
 * ones do, which is the only property that makes a stub rehearsal meaningful.
 */
function resolveProvider(args) {
  // opentable is network-blocked from this environment (Akamai edge denial) —
  // resy is the only reachable provider as of 2026-08-02. See providers/index.js.
  const name = args.provider ?? "resy";
  getProvider(name); // throws with the list of known providers
  return name;
}

const PAGE_ACTIONS = new Set(["restaurant.search", "restaurant.book"]);

async function handle(request) {
  const { action, arguments: args = {}, confirmed = false } = request;
  if (!action) throw new Error("request is missing `action`");

  if (IRREVERSIBLE.has(action) && confirmed !== true) {
    throw new Error(
      `refused: "${action}" is irreversible and arrived without upstream ` +
        "confirmation. The spoken gate in RunbookExecutor is authoritative; " +
        "this endpoint will not book on its own."
    );
  }

  // Real (non-stub) search/book need the live browser, which only exists
  // inside the Sailbox. Hop over once, here, rather than threading that
  // decision through every case below.
  if (!isStubMode() && !INSIDE_SAILBOX && PAGE_ACTIONS.has(action)) {
    return delegateToSailbox(request);
  }

  switch (action) {
    case "restaurant.search": {
      const params = toBookingParams(args);
      const provider = resolveProvider(args);
      if (isStubMode()) {
        return { candidates: 1, stub: true, query: params.restaurant, provider };
      }
      const page = await acquirePage();
      const results = await getProvider(provider).search(page, params);
      return { candidates: results.length, query: params.restaurant, provider };
    }

    case "restaurant.book": {
      const params = toBookingParams(args);
      const provider = resolveProvider(args);
      const page = isStubMode() ? undefined : await acquirePage();
      const booking = await bookStep({
        provider,
        params,
        page,
        // Already decided upstream, by voice. See CONFIRMATION POLICY above.
        getYes: async () => "yes",
        storePath: process.env.BOOKING_STORE_PATH,
      });
      return {
        confirmation_id: booking.confirmationRef,
        provider: booking.provider,
        stub: booking.stub === true,
      };
    }

    case "booking.reset": {
      const result = await resetAll({ storePath: process.env.BOOKING_STORE_PATH });
      return { cancelled: result.cancelled };
    }

    case "booking.list_open": {
      const open = await listOpenBookings(process.env.BOOKING_STORE_PATH);
      return { open: open.map((b) => b.confirmationRef) };
    }

    default:
      throw new Error(`unknown action "${action}"`);
  }
}

async function readStdin() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  return Buffer.concat(chunks).toString("utf8");
}

const raw = await readStdin();
try {
  const result = await handle(JSON.parse(raw));
  process.stdout.write(JSON.stringify({ ok: true, result }) + "\n");
} catch (error) {
  process.stdout.write(
    JSON.stringify({
      ok: false,
      error: error?.message ?? String(error),
      type: error?.constructor?.name ?? "Error",
    }) + "\n"
  );
  process.exitCode = 1;
}
